pkg/sql/opt/exec/execbuilder/testdata/upsert

```diff
@@ -670,3 +670,67 @@ vectorized: true
                           estimated row count: 333 (missing stats)
                           table: source@source_y_z_idx
                           spans: /NULL-/!NULL /1-
+
+# Test to make sure hash sharded index is used for uniqueness check
```

Comment:
I've realized there are quite a few interesting cases that we'll want to test to make sure we avoid uniqueness checks. I think you should move this test to `pkg/sql/opt/exec/execbuilder/testdata/hash_sharded_index` and add tests for all types of mutations including `INSERT`, `UPDATE`, `UPSERT`, `INSERT .. ON CONFLICT DO NOTHING`, and `INSERT .. ON CONFLICT .. DO UPDATE`.